### PR TITLE
Implement Weak::new_downgraded() (#30425)

### PR DESCRIPTION
**Pull request by shahn**
This adds a constructor for a Weak that can never be upgraded. These are
mostly useless, but for example are required when deserializing.

### CODE REVIEW

**Comment by brson** [comment]
cc @rust-lang/libs Is everyone cool with this? Does it need an RFC?

File: src/liballoc/rc.rs [hunk 1]
```diff
@@ -830,6 +830,36 @@ impl<T: ?Sized + fmt::Debug> fmt::Debug for Weak<T> {
...
+    #[unstable(feature = "downgraded_weak",
+               reason = "recently added",
+               issue="30425")]
+    pub fn new_downgraded() -> Weak<T> {
+        unsafe {
+            Weak {
+                _ptr: Shared::new(Box::into_raw(box RcBox {
+                    strong: Cell::new(0),
+                    weak: Cell::new(1),
+                    value: uninit(),
```
Comment (apasel422) [inline src/liballoc/rc.rs]: Can this use `core::mem::uninitialized` instead?

**Comment by alexcrichton** [comment]
I would probably personally call this `Weak::new` to stick closer to our existing conventions, but otherwise it seems harmless to me as well.

**Comment by apasel422** [comment]
Should `Arc`'s `Weak` have this as well?

File: src/liballoc/rc.rs (hunk 1 above)
Comment (shahn) [inline src/liballoc/rc.rs]: fixing

**Comment by shahn** [comment]
@alexcrichton The reason I went with a different name was to make it clear that this won't be usable for "normal" code. Happy to defer to you of course, extending PR.

@apasei422 good point! I'll extend the PR.

**Comment by shahn** [comment]
I'm thinking if everyone likes this branch I should probably rebase & squash it into a single commit?

File: src/liballoc/arc.rs [hunk 2]
```diff
@@ -910,6 +911,36 @@ impl<T> From<T> for Arc<T> {
...
+    /// Weak<T>::upgrade() on the return value always gives None.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(downgraded_weak)]
+    ///
+    /// use std::sync::Arc;
+    ///
+    /// let five = Arc::new(5);
```
Comment (apasel422) [inline src/liballoc/arc.rs]: This should be `Weak::new`.

File: src/liballoc/rc.rs [hunk 3]
```diff
@@ -830,6 +830,38 @@ impl<T: ?Sized + fmt::Debug> fmt::Debug for Weak<T> {
...
+    /// Weak<T>::upgrade() on the return value always gives None.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(downgraded_weak)]
+    ///
+    /// use std::rc::Weak;
+    ///
+    /// let empty:Weak<i64> = Weak::new();
```
Comment (apasel422) [inline src/liballoc/rc.rs]: There should be a space before `Weak<i64>`.

File: src/liballoc/arc.rs [hunk 4]
```diff
@@ -910,6 +911,36 @@ impl<T> From<T> for Arc<T> {
...
+    /// ```
+    #[unstable(feature = "downgraded_weak",
+               reason = "recently added",
+               issue = "30425")]
+    pub fn new() -> Weak<T> {
+        unsafe {
+            let x: Box<_> = box ArcInner {
```
Comment (apasel422) [inline src/liballoc/arc.rs]: Is there a reason this `Box` is assigned to a local variable when the one for `rc::Weak` isn't?

Comment (shahn) [inline src/liballoc/arc.rs]: I mirrored Arc::new(). But I can remove it.

**Comment by shahn** [comment]
I agree wrt the examples, I added them because they were there for Rc::new and Arc::new (where they also only assign a value). Should I update this?

**Comment by shahn** [comment]
I think I addressed all review comments, thanks a  bunch. Anything else for me to do except wait while a decision is made?

**Pull request merged**